% Inherited from the classic.tplx https://pypi.org/project/nb-pdf-template/#description %
\documentclass[11pt]{article}

    \usepackage{parskip}
    \setcounter{secnumdepth}{0} %Suppress section numbers
    \usepackage[breakable]{tcolorbox}
    \tcbset{nobeforeafter}
    \usepackage{needspace}
    
    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}
    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range
    \makeatletter % fix for grffile with XeLaTeX
    \def\Gread@@xetex#1{%
      \IfFileExists{"\Gin@base".bb}%
      {\Gread@eps{\Gin@base.bb}}%
      {\Gread@@xetex@aux#1}%
    }
    \makeatother

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}
    

    \let\Oldtex\TeX     % provide compatibility with nbconvert <= 5.3.1
    \let\Oldlatex\LaTeX % pre-included in nbconvert > 5.3.1 so redundant
    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    
    
    
    
% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother

    %Reconfigured pygments
    \makeatletter
    \expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[HTML]{008800}{##1}}} %numbers
    \expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[HTML]{008800}{##1}}} %numbers
    \expandafter\def\csname PY@tok@nn\endcsname{\def\PY@tc##1{\textcolor[HTML]{000000}{##1}}} %imports
    \expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[HTML]{008000}{##1}}} %operator.word
    \expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[HTML]{AA22FF}{\codetrue##1\codefalse}}} %operator
    \makeatother

    \makeatletter
    \newcommand*\@iflatexlater{\@ifl@t@r\fmtversion}
    \@iflatexlater{2016/03/01}{
	    \newcommand{\wordboundary}{4095}}{
	    \newcommand{\wordboundary}{255}}
    \makeatother

    \newif\ifcode
    \codefalse
    \definecolor{Grey}{rgb}{0.40,0.40,0.40}
    %If using XeLaTeX, use magic to not highlight . operators with purple.
    \ifdefined\XeTeXcharclass
    \XeTeXinterchartokenstate = 1
    \newXeTeXintercharclass \mycharclassGrey
    \XeTeXcharclass `. \mycharclassGrey
    \XeTeXinterchartoks 0 \mycharclassGrey   = {\bgroup\ifcode\color{Grey}\else\fi}

    \XeTeXinterchartoks \wordboundary \mycharclassGrey = {\bgroup\ifcode\color{Grey}\else\fi}

    \XeTeXinterchartoks \mycharclassGrey 0   = {\egroup}
    \XeTeXinterchartoks \mycharclassGrey \wordboundary = {\egroup}
    \fi %end magical operator highlighting
    %End Reconfigured Pygments
    
   
    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}

    % needed definitions
    \newif\ifleftmargins
    \newlength{\promptlength}

    % cell style settings
        \leftmarginsfalse

    
    % prompt
    \newcommand{\prompt}[3]{
        \needspace{1.1cm}
        \settowidth{\promptlength}{ #1 [#3] }
        \ifleftmargins\hspace{-\promptlength}\hspace{-5pt}\fi
        {\color{#2}#1 [#3]:}
        \ifleftmargins\vspace{-2.7ex}\fi
    }
    
    
    % environments
    \newenvironment{InVerbatim}{\VerbatimEnvironment%
        \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,
            colback=cellbackground, colframe=cellborder]
            \begin{Verbatim}
            }{
            \end{Verbatim}
        \end{tcolorbox}
    }
    \newenvironment{OutVerbatim}{\VerbatimEnvironment%
        \begin{tcolorbox}[breakable, boxrule=.5pt, size=fbox, pad at break*=1mm, opacityfill=0]
            \begin{Verbatim}
            }{
            \end{Verbatim}
        \end{tcolorbox}
    }
    
    %Updated MathJax Compatibility (if future behaviour of the notebook changes this may be removed)
    \renewcommand{\TeX}{\ifmmode \textrm{\Oldtex} \else \textbackslash TeX \fi}
    \renewcommand{\LaTeX}{\ifmmode \Oldlatex \else \textbackslash LaTeX \fi}
    
    % Header Adjustments
    \renewcommand{\paragraph}{\textbf}
    \renewcommand{\subparagraph}[1]{\textit{\textbf{#1}}}

    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    \geometry{verbose,tmargin=.5in,bmargin=.7in,lmargin=.5in,rmargin=.5in}
    

\begin{document}
    
    \title{Linking TCGA's to Lung Cancers\\Clustering Lung Cancer Patients by Genomic Information}\author{Haoda Li}


\date{\today}
\maketitle


    
    

    

    \hypertarget{summary}{%
\section{\texorpdfstring{\emph{Summary}}{Summary}}\label{summary}}

\emph{This report examines whether lung cancer patients' with different
survival time is correlated with distinct genomic subgroups. Three
distinct clusters are formed through dimension reduction and community
finding and showed different survival curves. The report further
considers the the association among different observed clinical factors
and their genomic features, while the associations were not found for
gender, age, and cancer's tumor stage. This report concluds that the
genomic subgroups can be used to predict the patients' survival curve.}

    \hypertarget{introduction}{%
\section{1. Introduction}\label{introduction}}

Human genes are sequences of tiny piece chemicals that are often encoded
as T, C, G, and A's. With the examinations on such sequences of letters,
the scientists are able to find clues of how cancers happen. However,
the gene sequences are as long as millions of letters, hence are almost
unable to interpret. One possible solution is to reduce the dimensions
and build clusters so that each cluster show different biological
features.

The report will focus on 424 lung cancer patients and their 60483 gene
sequence expressions from TCGA \cite{weinstein2013cancer} and find
distinct gene subgroups that have different survival curves. In
addition, clinical features such as age, gender, and tumor stage and
examined. We aim to find whether the gene group plays a unique role in
predicting a patient's survival.


    \hypertarget{methodology}{%
\section{2. Methodology}\label{methodology}}

Overall, the data is divided into 2 parts. The gene scanning and the
clinical data. For the gene data, because of the extremely large
dimensionality, we need to first reduce the dimension. Then, we choose
clusters from different clustering algorithms and evaluate how well the
clustering performs on the patients' survival data. Afterwards, other
clinical features are then incorporated and we try to build regression
models that can explain the relationship among different features and
the survival time.

    \hypertarget{clustering-on-gene-data}{%
\subsection{2.1. Clustering on Gene
Data}\label{clustering-on-gene-data}}

\hypertarget{preprocessing}{%
\subsubsection{2.1.1.Preprocessing}\label{preprocessing}}

The first difficulty is that the data points positive and extremely
right skewed. As shown in the plot, half of the data points are close to
0 and the other half are arbitrarily large. Therefore, we normalize the
data points and take a log one plus transformation so that the data.
After the normalization, the data is less right skewed and numerically
smaller, which makes further processing easier.


    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_6_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    Also, as shown in Fig 2, note that the data is close to spare. 80\% of
the data entries is close to zero. Therefore, we can use principle
component analysis to reduce the dimension. PCA can then preserve the
most variations and the lower dimensions is easier to be interpreted and
analyzed. After performing PCA, we take the first 50 components, which
preserves the majority of the variance. As shown in Fig 2, the data
entries are dense and the first few dimensions have much variations.


    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_8_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \hypertarget{clustering}{%
\subsubsection{2.1.2. Clustering}\label{clustering}}

Among many clustering algorithms, we choose to use Louvain algorithm.
Louvain algorithm \cite{blondel2008fast} are adapted from network
theorem and has great performance on clustering gene expressions.
Louvain algorithm works by first constructing a neighborhood graph ,
which is to assign edges based on Euclidean distance among genes. Then,
Louvain algorithm tends to find ``communities'', which are subgroups of
genes that are closed and have many connections in the neighborhood
graph. Compare to distance based clustering algorithm such as k-nearest
neighbors, Louvain algorithm tends to capture the internal connections
based on the number of points within some distance. Rather than finding
hyper-balls shaped clusters by kNN, Louvain algorithm can find various
clusters with different shapes.

    \hypertarget{other-factors-from-clinical-data}{%
\subsection{2.2. Other Factors from Clinical
Data}\label{other-factors-from-clinical-data}}





    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_14_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    Taking consideration of other clinical factors, including age at the
diagnosis, tumor stage, and gender. Naturally, the aged people are more
likely to be diagnosed with cancers. Which is reflected in our dataset,
as the majority of the patients are centered around 60. However, due to
the low variations of age in our dataset, evaluating the relationship
between age and survival time is hard. In addition, tumor stages
definitely have decisive impact on the survival time, as shown in Fig 5.
Finally, there is no evidence that gender is related to survival, as
shown in the survival curve. Also, note that the number of observations
between

    \hypertarget{models}{%
\subsection{2.3. Models}\label{models}}

From the investigations above, two Cox regression models are proposed.
The additive model considers the effect of gene clusters, age, and tumor
stage on survival time. The interaction model, in addition, consider the
interactions of gene clusters vs.~age and gene clusters vs.~tumor stage.
The models are given as
\[h_1 = \text{baseline} \cdot \exp(\beta_1\overline{\text{cluster}} + \beta_2\overline{\text{tumor}} + \beta_3\overline{\text{age}})\]
\[h_2 = \text{baseline} \cdot \exp(\beta_1\overline{\text{cluster}} + \beta_2\overline{\text{tumor}} + \beta_3\overline{\text{age}} + \beta_4\overline{\text{cluster:tumor}} + \beta_5\overline{\text{cluster:age}})\]
where \(\overline{\text{group}}\) are zero-centered observations. We use
the interactive model to assess whether the gene clusters is correlated
with age or tumor stage. On the other words, whether the gene groups is
a unique predictor that can provide additional information than clinical
observations.

    \hypertarget{results}{%
\section{3. Results}\label{results}}

    The clusters of genes are plotted on a 2D UMAP\cite{lel2018umap} basis.
UMAP is a non-linear dimension reduction algorithm that approximates a
manifold from the neighborhood graph. UMAP tends to give a more
intuitive visualization of the distance and connections. We also show
the dendrogram to see the hierarchical clusters.


    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_19_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    The performance of our clusters on survival time is evaluated through
Kaplan-Meier estimate. However, from the plot, we see some of the
clusters from the Louvain algorithm have very similar similar survival
curve, hence we will use 3 clusters from hierarchical clusters at level
2 of the dendrogram.


    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_21_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    Then, we fit the regression models based on the clusters and the
coefficients of the fitted models are shown below.





    \begin{tabular}{lllll}
\toprule
                      name & coef. of model1 & p-value of model1 & coef. of model2 & p-value of model2 \\
\midrule
               louvain\_h=2 &       -0.489939 &         0.0387053 &        -1.77848 &          0.145783 \\
               louvain\_h=3 &       -0.568266 &         0.0148153 &        -1.19727 &          0.271095 \\
             tumor\_stage=2 &        0.359658 &          0.153736 &         0.78168 &          0.110364 \\
             tumor\_stage=3 &        0.990963 &       4.89213e-06 &         1.55286 &       7.68451e-05 \\
             tumor\_stage=4 &         1.85683 &        0.00215931 &         1.84658 &         0.0138706 \\
              age\_at\_index &       0.0133238 &         0.0740811 &     -0.00365337 &          0.772061 \\
 louvain\_h=2:tumor\_stage=2 &                 &                   &       -0.373705 &           0.56589 \\
 louvain\_h=3:tumor\_stage=2 &                 &                   &       -0.737301 &          0.246352 \\
 louvain\_h=2:tumor\_stage=3 &                 &                   &       -0.922433 &         0.0938969 \\
 louvain\_h=3:tumor\_stage=3 &                 &                   &       -0.603801 &          0.261471 \\
 louvain\_h=2:tumor\_stage=4 &                 &                   &       -0.241636 &          0.850306 \\
 louvain\_h=3:tumor\_stage=4 &                 &                   &                 &                   \\
  louvain\_h=2:age\_at\_index &                 &                   &       0.0294332 &          0.131746 \\
  louvain\_h=3:age\_at\_index &                 &                   &       0.0179933 &           0.30297 \\
\bottomrule
\end{tabular}


    
    The p-values of all interaction terms are extremely large. In addition,
the overall p-values in model 2 are much larger than model 1. This
observation hints multicollinearity problem in model 2. Therefore, we
conclude that there is no clues of interactions for cluster vs.~tumor
stage and cluster vs.~age. In addition, the independence of cluster
vs.~age and cluster vs.~age is hinted in Fig 6, where the distributions
are similar among cluster groups.

For coefficients in model 1, there is some evidence that the clusters is
associated with survival time. Although this association is much weaker
than tumor stage. Because the p-value is still close to 0.05 and the
model does not fit quite well, this clustering may still be
insignificant.


    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{report_files/report_28_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \hypertarget{conclusion}{%
\section{4. Conclusion}\label{conclusion}}

    In conclusion, the gene subgroups that found by PCA dimension reduction
and Louvain algorithm provide some clues of survival times for the
patients with lung cancer. In addition, the gene clustering information
is likely to be independent of the observed clinical information, such
as age, gender, and tumor stage. Therefore, measuring the genes of the
patients can provide unique information for the treatment plan.

However, we have to notice that the Cox regression fit is not quite
good, and the coefficients of the clusters are just below 0.05. Because
the dimensionality of genes is much greater than the number of
observations, this model may not be generalized to other cancer cases.
Further validations and testings should be conducted on other cases to
measure the performance of this clustering procedure. In addition, these
gene clusters can correlate with some clinical factors that is not
examined, and the clusters may not have unique and independent impact on
the severity of the cancer. We need to consider more clinical factors
that may be correlated with the gene clusters.

    \bibliographystyle{unsrt}
\bibliography{ref}


    % Add a bibliography block to the postdoc
    
    
    
\end{document}
